rclcpp/include/rclcpp/wait_set_policies/dynamic_storage.hpp

```diff
@@ -0,0 +1,161 @@
+class DynamicStorage : public rclcpp::wait_set_policies::detail::StoragePolicyCommon<false>
+{
+protected:
+  using is_mutable = std::true_type;
+
+  using SequenceOfWeakGuardConditions = std::vector<std::weak_ptr<rclcpp::GuardCondition>>;
+  using GuardConditionsIterable = std::vector<std::shared_ptr<rclcpp::GuardCondition>>;
+
+  explicit
+  DynamicStorage(
+    const GuardConditionsIterable & guard_conditions,
+    rclcpp::Context::SharedPtr context
+  )
+  : StoragePolicyCommon(guard_conditions, context),
+    guard_conditions_(guard_conditions.cbegin(), guard_conditions.cend()),
+    shared_guard_conditions_(guard_conditions_.size())
+  {}
+
+  ~DynamicStorage() = default;
+
+  void
+  storage_rebuild_rcl_wait_set()
+  {
+    this->storage_rebuild_rcl_wait_set_with_sets(
+      guard_conditions_
+    );
+  }
+
+  bool
+  storage_has_guard_condition(const rclcpp::GuardCondition & guard_condition)
+  {
+    return std::any_of(
+      guard_conditions_.cbegin(),
+      guard_conditions_.cend(),
+      [&guard_condition](const auto & gc) { return &guard_condition == gc.lock().get(); });
+  }
+
+  auto
+  storage_find_guard_condition(const rclcpp::GuardCondition & guard_condition)
+  {
+    return std::find_if(
+      guard_conditions_.begin(),
+      guard_conditions_.end(),
+      [&guard_condition](const auto & gc) { return &guard_condition == gc.lock().get(); });
+  }
+
+  void
+  storage_add_guard_condition(std::shared_ptr<rclcpp::GuardCondition> && guard_condition)
+  {
+    if (this->storage_has_guard_condition(*guard_condition)) {
+      throw std::runtime_error("guard_condition already in wait set");
+    }
+    guard_conditions_.push_back(std::move(guard_condition));
+    this->storage_flag_for_resize();
+  }
+
+  void
+  storage_remove_guard_condition(std::shared_ptr<rclcpp::GuardCondition> && guard_condition)
+  {
+    auto it = this->storage_find_guard_condition(*guard_condition);
+    if (guard_conditions_.cend() == it) {
+      throw std::runtime_error("guard_condition not in wait set");
+    }
+    guard_conditions_.erase(it);
+    this->storage_flag_for_resize();
+  }
+
+  // this is noexcept because:
+  //   - std::weak_ptr::expired is noexcept
+  //   - the erase-remove idiom is noexcept, since we're not using the ExecutionPolicy version
+  //   - std::vector::erase is noexcept if the assignment operator of T is also
+  //     - and, the operator= for std::weak_ptr is noexcept
+  void
+  storage_prune_deleted_entities() noexcept
+  {
+    // reusable (templated) lambda for removal predicate
+    auto p =
+      [](const auto & weak_ptr) {
+        // remove entries which have expired
+        return weak_ptr.expired();
+      };
+    // remove guard conditions which have been deleted
+    guard_conditions_.erase(std::remove_if(guard_conditions_.begin(), guard_conditions_.end(), p));
+  }
+
+  void
+  storage_acquire_ownerships()
+  {
+    if (++ownership_reference_counter_ > 1) {
```

Comment:
If `ownership_reference_counter_` is `0`, then after `++` it will be `1`, which is `== 1`, but not `> 1`. I think it makes sense, but please check my logic.

Reply:
Makes sense, I must have had `>= 1` in my mind.